### ISSUE TITLE
unclear issue with columns not uniquely identified

### ISSUE THREAD

**Issue by sbfnk** [epiforecasts/qra#2]
see #1 

@nikosbosse can you create a reproducible example, perhaps based on the example in the [README](https://github.com/epiforecasts/qra/blob/master/README.md)?

**Comment by seabbs** [comment]
@nikosbosse I think this may be an issue with the input data not the QRA package. `spread` throws errors like this when a variable it's not expecting to be there is present.

**Comment by nikosbosse** [comment]
Agree the input data seems to play a role. But: I tried debugging and even tried spreading manually (used pivot_wider) and manually specified all columns - the error persisted and only disappeared when I removed non-unique rows. 

The fact that we get non-unique rows comes from the input I guess

Will work on a reprex to see what is happening

**Comment by seabbs** [comment]
I'm not clear why non-unique rows are present in the input data? It can't handle that anywhere.

**Comment by nikosbosse** [comment]
there aren't any non-unique rows in the input data. but somewhere non-unique rows appear in the step before spreading.

**Comment by sbfnk** [comment]
The more general issue is that it might be good at some point to split out the core functionality to work with a simpler data structure rather than requring data frames with specific columns.

**Comment by nikosbosse** [comment]
I was able to "fix" the issue with actual data, but wasn't able to produce a reprex. 
The problem went away when deleting a column with state codes that was present in addition to identifying locations by name. So there should be no new information in that column and the code ran after deleting that column. 

Artifically adding a location code column in the Readme data however wasn't able to break the code. I added another column like this
```
# data and forecasts generated as in the Readme
code <- data.frame(geography = c("region 1", "region 2", "region 3", "country"),
                   code = c(1, 2, 3, 4))

fc <- forecasts %>%
  inner_join(code)

res <- qra::qra(fc, data, pool = c("horizon", "geography"),
                min_date = max(forecasts$creation_date) - 13)
```
but that worked without a problem. 
So I am still not sure what causes the problem and how to really fix it. 
Digging deeper...

**Comment by nikosbosse** [comment]
My journey down the rabbit hole. 
TLDR: 
- I would suggest dropping the location column when doing the qra analysis. If my understanding is correct this means that we will not have weights that differ by state, but instead we have one weight for the region scale and one weight for the country scale. 
- merging the PR should effectively prevent the code from returning errors in the future, but I am concerned it could mask underlying problems. I still haven't understood where the issue comes from and would therefore suggest not to merge it. 

Here is the longer take: 
ok so the first difference first appears here in the code: 
![image](https://user-images.githubusercontent.com/37978797/86246379-3159ec80-bbab-11ea-8927-aad5337b5008.png)

if we have locations included, then max_horizons will be per location 
![image](https://user-images.githubusercontent.com/37978797/86246436-46368000-bbab-11ea-8f64-c691862a78bf.png)

whereas if location is not included, max_horizons will be only one per geography sclae
![image](https://user-images.githubusercontent.com/37978797/86246311-14bdb480-bbab-11ea-968f-b2d5bdf447df.png)

But I assume that is fine - it works the same way with the example data in the readme. 

I think the real magic happens here: 
![image](https://user-images.githubusercontent.com/37978797/86249055-08d3f180-bbaf-11ea-942c-b64a0d137019.png)
Unfortunately I don't quite understand what the code does. The result though is this: 
with locations given, `complete_dataset` has more rows. filtering for Alabama, I notice that with locations present, there are 5 days ahead and 12 days ahead forecasts, whereas without locations given there are only 5 day ahead forecasts and the others are missing. 

With the nesting in the next step we either have two categories with the same amount of data
![image](https://user-images.githubusercontent.com/37978797/86249545-b7783200-bbaf-11ea-87cf-d663c57339f0.png)
or more with differing amounts of data
![image](https://user-images.githubusercontent.com/37978797/86249638-db3b7800-bbaf-11ea-97b8-0227f99b2322.png)

Getting the ensemble works in both cases. However: if we ensemble by locations, there are 8,890 rows. If we ensemble without location specific weights, there are only 8,250 rows. Don't know where the difference comes from. Values in the location included case looked a bit weird. 

The error then appears here: 
![image](https://user-images.githubusercontent.com/37978797/86250952-8bf64700-bbb1-11ea-8f94-5092ec933d50.png)

**Comment by sbfnk** [comment]
Have you tried the `pool` option? If you specify, e.g., `pool=c("horizon", "location")` it should not try to estimate location-specific weights. In general the code treats any column it doesn't recognise as a relevant column for which to estimate separate weights, unless specified with `pool`.

**Comment by nikosbosse** [comment]
running 
```
res <- qra::qra(forecasts_for_qra, death_data, pool = c("horizon", "geography", "location"),
                min_date = max(forecasts_for_qra$creation_date) - 20, 
                enforce_normalisation = TRUE)
```
gives an empty data.frame --> I assume because geography and location basically is the same?

**Comment by sbfnk** [comment]
Yes, the function assumes that every combination of columns is present in the data (except those that have special meaning, e.g. "value"). Because of this you can't have redundant columns, one of "location" or "geography" would have to go.
